File: 981-time-based-key-value-store.py
class TimeMap(object):

    def __init__(self):
        self.mapping = defaultdict(list)

    def set(self, key, value, timestamp):
        """
        :type key: str
        :type value: str
        :type timestamp: int
        :rtype: None
        """
        self.mapping[key].append((timestamp, value))
    def get(self, key, timestamp):
        """
        :type key: str
        :type timestamp: int
        :rtype: str
        """
        if not key in self.mapping or len(self.mapping[key]) == 0:
            return ""
        stampList = self.mapping[key]
        left = 0
        right = len(stampList) - 1
        while left < right:
            mid = int((left + right) / 2)
            if stampList[mid][0] == timestamp:
                return stampList[mid][1]
            elif stampList[mid][0] < timestamp:
                if mid != len(stampList) - 1 and stampList[mid + 1][0] > timestamp:
                    return stampList[mid][1]
                left = mid + 1 
            else:
                right = mid - 1
        if left == 0 and stampList[left][0] > timestamp:
            return ""
        return stampList[left][1]



# Your TimeMap object will be instantiated and called as such:
# obj = TimeMap()
# obj.set(key,value,timestamp)
# param_2 = obj.get(key,timestamp)

File: 981-time-based-key-value-store.cpp
class TimeMap {
public:
    unordered_map<string, std::set<pair<int, string>>> mapping;
    TimeMap() {
        // unordered_map<string, std::set<pair<int, int>>> mapping;
    }
    
    void set(string key, string value, int timestamp) {
        mapping[key].insert(pair(timestamp, value));
    }
    
    string get(string key, int timestamp) {
        auto& s = mapping[key];
        if (s.empty()) {
            return "";
        }
        auto it = s.upper_bound({timestamp, string(1, CHAR_MAX)});
        if (it == s.begin()) {
            return "";
        }
        return prev(it)->second;
    }
};

/**
 * Your TimeMap object will be instantiated and called as such:
 * TimeMap* obj = new TimeMap();
 * obj->set(key,value,timestamp);
 * string param_2 = obj->get(key,timestamp);
 */